src/Renci.SshNet/ForwardedPortDynamic.cs

```diff
@@ -112,51 +125,620 @@ protected override void StopPort(TimeSpan timeout)
+            // and at that time there may still be pending requests
+            var pendingChannelCountdown = _pendingChannelCountdown;
+
+            pendingChannelCountdown.AddCount();
+
+            try
+            {
+                using (var channel = Session.CreateChannelDirectTcpip())
+                {
+                    channel.Exception += Channel_Exception;
+
+                    if (!HandleSocks(channel, clientSocket, Session.ConnectionInfo.Timeout))
+                    {
+                        CloseClientSocket(clientSocket);
+                        return;
+                    }
+
+                    // start receiving from client socket (and sending to server)
+                    channel.Bind();
+                }
+            }
+            catch (Exception exp)
+            {
+                RaiseExceptionEvent(exp);
+                CloseClientSocket(clientSocket);
+            }
+            finally
+            {
+                // take into account that CountdownEvent has since been disposed; when stopping the port we
+                // wait for a given time for the channels to close, but once that timeout period has elapsed
+                // the CountdownEvent will be disposed
+                try
+                {
+                    _ = pendingChannelCountdown.Signal();
+                }
+                catch (ObjectDisposedException)
+                {
+                    // Ignore any ObjectDisposedException
+                }
+            }
+        }
 
         /// <summary>
-        /// Stops the listener.
+        /// Initializes the <see cref="CountdownEvent"/>.
         /// </summary>
-        partial void StopListener();
+        /// <remarks>
+        /// <para>
+        /// When the port is started for the first time, a <see cref="CountdownEvent"/> is created with an initial count
+        /// of <c>1</c>.
+        /// </para>
+        /// <para>
+        /// On subsequent (re)starts, we'll dispose the current <see cref="CountdownEvent"/> and create a new one with
+        /// initial count of <c>1</c>.
+        /// </para>
+        /// </remarks>
+        private void InitializePendingChannelCountdown()
+        {
+            var original = Interlocked.Exchange(ref _pendingChannelCountdown, new CountdownEvent(1));
+            original?.Dispose();
+        }
+
+        private bool HandleSocks(IChannelDirectTcpip channel, Socket clientSocket, TimeSpan timeout)
+        {
+#pragma warning disable IDE0039 // Use lambda instead of local function to reduce allocations
+            // Create eventhandler which is to be invoked to interrupt a blocking receive
+            // when we're closing the forwarded port.
+            EventHandler closeClientSocket = (_, args) => CloseClientSocket(clientSocket);
+#pragma warning restore IDE0039 // Use lambda instead of local function to reduce allocations
```

Comment:
The suppression confuses me... it is already a lambda?

Reply:
If I'm not mistaken, I didn't add this suppression. It was probably already in the partial class that I merged.
I'll have a look though.

Reply:
Updated code to use local function instead. Removed suppression.